File: apps/ios/SignalASI/VoiceAudioHub.swift
import Foundation

protocol PcmRecorder {
  func start(config: PcmCaptureConfig) throws -> AsyncStream<AudioFrame>
  func requestStop(reason: PcmStopReason)
  func stop(reason: PcmStopReason) async
  func currentState() -> PcmRecorderState
}

struct VoiceAudioSessionConfig: Equatable {
  var capture: PcmCaptureConfig
  var endpoint: AdaptiveEndpointConfig
  var autoEndpoint: Bool

  init(
    capture: PcmCaptureConfig = PcmCaptureConfig(),
    endpoint: AdaptiveEndpointConfig = AdaptiveEndpointConfig(),
    autoEndpoint: Bool = false
  ) {
    self.capture = capture
    self.endpoint = endpoint
    self.autoEndpoint = autoEndpoint
  }
}

struct VoiceAudioSession: Equatable {
  var id: String
}

protocol VoiceAudioHubListener {
  func onCaptureReady(session: VoiceAudioSession, state: PcmRecorderState)
  func onAudioLevel(session: VoiceAudioSession, decision: VadDecision)
  func onSpeechStarted(session: VoiceAudioSession, sequence: Int64)
  func onSpeechEndedCandidate(session: VoiceAudioSession, sequence: Int64)
  func onEndpoint(session: VoiceAudioSession, reason: EndpointReason)
  func onInputRouteChanged(session: VoiceAudioSession, route: String)
  func onFailure(session: VoiceAudioSession, error: Error)
}

extension VoiceAudioHubListener {
  func onCaptureReady(session: VoiceAudioSession, state: PcmRecorderState) {}
  func onAudioLevel(session: VoiceAudioSession, decision: VadDecision) {}
  func onSpeechStarted(session: VoiceAudioSession, sequence: Int64) {}
  func onSpeechEndedCandidate(session: VoiceAudioSession, sequence: Int64) {}
  func onEndpoint(session: VoiceAudioSession, reason: EndpointReason) {}
  func onInputRouteChanged(session: VoiceAudioSession, route: String) {}
  func onFailure(session: VoiceAudioSession, error: Error) {}
}

struct NoopVoiceAudioHubListener: VoiceAudioHubListener {}

final class VoiceAudioHub {
  private final class ActiveSession {
    let publicSession: VoiceAudioSession
    let config: VoiceAudioSessionConfig
    let store: SpeechSegmentStore
    let vad: VoiceActivityDetector
    let endpoint: AdaptiveEndpointDetector
    let listener: VoiceAudioHubListener
    let lock = NSLock()
    var stopRequested = false
    var lastRoute = ""
    var task: Task<Void, Never>?

    init(
      publicSession: VoiceAudioSession,
      config: VoiceAudioSessionConfig,
      store: SpeechSegmentStore,
      vad: VoiceActivityDetector,
      endpoint: AdaptiveEndpointDetector,
      listener: VoiceAudioHubListener
    ) {
      self.publicSession = publicSession
      self.config = config
      self.store = store
      self.vad = vad
      self.endpoint = endpoint
      self.listener = listener
    }

    func isStopRequested() -> Bool {
      locked { stopRequested }
    }

    func requestStop() {
      locked { stopRequested = true }
    }

    func claimEndpointStop() -> Bool {
      locked {
        guard !stopRequested else { return false }
        stopRequested = true
        return true
      }
    }

    func updateRoute(_ route: String) -> Bool {
      locked {
        guard !route.isEmpty, route != lastRoute else { return false }
        lastRoute = route
        return true
      }
    }

    private func locked<T>(_ action: () -> T) -> T {
      lock.lock()
      defer { lock.unlock() }
      return action()
    }
  }

  private let recorder: PcmRecorder
  private let sessionIdFactory: () -> String
  private let vadFactory: () -> VoiceActivityDetector
  private let lock = NSLock()
  private var active: ActiveSession?

  init(
    recorder: PcmRecorder,
    sessionIdFactory: @escaping () -> String = { UUID().uuidString },
    vadFactory: @escaping () -> VoiceActivityDetector = { AdaptiveSpeechVad() }
  ) {
    self.recorder = recorder
    self.sessionIdFactory = sessionIdFactory
    self.vadFactory = vadFactory
  }

  func start(
    config: VoiceAudioSessionConfig,
    listener: VoiceAudioHubListener = NoopVoiceAudioHubListener()
  ) -> VoiceAudioSession? {
    let session = locked { () -> ActiveSession? in
      guard active == nil else { return nil }
      let publicSession = VoiceAudioSession(id: sessionIdFactory())
      let session = ActiveSession(
        publicSession: publicSession,
        config: config,
        store: InMemorySpeechSegmentStore(
          sampleRateHz: config.capture.sampleRateHz,
          maxDurationMs: config.capture.maxDurationMs + 1_000
        ),
        vad: vadFactory(),
        endpoint: AdaptiveEndpointDetector(
          sampleRateHz: config.capture.sampleRateHz,
          config: config.endpoint,
          autoEndpoint: config.autoEndpoint
        ),
        listener: listener
      )
      active = session
      return session
    }
    guard let session = session else { return nil }
    session.task = Task { [weak self] in
      await self?.run(session)
    }
    return session.publicSession
  }

  func currentAmplitude() -> Int {
    recorder.currentState().currentAmplitude
  }

  func currentState() -> PcmRecorderState {
    recorder.currentState()
  }

  func requestStop(session: VoiceAudioSession, reason: PcmStopReason) {
    guard let current = matchingActive(session) else { return }
    current.requestStop()
    recorder.requestStop(reason: reason)
  }

  func stop(session: VoiceAudioSession, reason: PcmStopReason) async -> VoiceAudioCaptureResult? {
    guard let current = matchingActive(session) else { return nil }
    current.requestStop()
    await recorder.stop(reason: reason)
    await current.task?.value
    let state = recorder.currentState()
    let result = VoiceAudioCaptureResult(
      sessionId: session.id,
      stopReason: state.stopReason ?? reason,
      snapshot: current.store.snapshot(
        segment: SegmentRange(
          preRollMs: current.config.endpoint.preRollMs,
          postRollMs: current.config.endpoint.postRollMs
        )
      ),
      diagnostics: state.diagnostics,
      audioSource: state.audioSource,
      inputRoute: state.inputRoute
    )
    current.store.clear()
    clearActive(current)
    return result
  }

  func activeSession() -> VoiceAudioSession? {
    locked { active?.publicSession }
  }

  func snapshotWindow(session: VoiceAudioSession, maxDurationMs: Int64) -> PcmSnapshot? {
    guard let current = matchingActive(session) else { return nil }
    return current.store.snapshotWindow(
      maxDurationMs: maxDurationMs,
      segment: SegmentRange(
        preRollMs: current.config.endpoint.preRollMs,
        postRollMs: current.config.endpoint.postRollMs
      )
    )
  }

  private func run(_ session: ActiveSession) async {
    do {
      let frames = try recorder.start(config: session.config.capture)
      if session.isStopRequested() {
        await recorder.stop(reason: .userCancel)
        return
      }
      let readyState = recorder.currentState()
      _ = session.updateRoute(readyState.inputRoute)
      session.listener.onCaptureReady(session: session.publicSession, state: readyState)
      for await frame in frames {
        processFrame(session, frame: frame)
        frame.close()
      }
    } catch {
      if !session.isStopRequested() {
        session.listener.onFailure(session: session.publicSession, error: error)
      }
    }
    if !session.isStopRequested() {
      session.store.clear()
      clearActive(session)
    }
  }

  private func processFrame(_ session: ActiveSession, frame: AudioFrame) {
    session.store.append(frame)
    let decision = session.vad.accept(frame)
    let endpoint = session.endpoint.accept(frame, vad: decision)
    if endpoint.speechStarted {
      session.store.markSpeechStart(sequence: frame.sequence)
      session.listener.onSpeechStarted(session: session.publicSession, sequence: frame.sequence)
    }
    if endpoint.speechEndedCandidate {
      session.store.markSpeechEnd(sequence: frame.sequence)
      session.listener.onSpeechEndedCandidate(session: session.publicSession, sequence: frame.sequence)
    }
    session.listener.onAudioLevel(session: session.publicSession, decision: decision)
    let route = recorder.currentState().inputRoute
    if session.updateRoute(route) {
      session.listener.onInputRouteChanged(session: session.publicSession, route: route)
    }
    if let reason = endpoint.endpointReason, session.claimEndpointStop() {
      session.listener.onEndpoint(session: session.publicSession, reason: reason)
    }
  }

  private func matchingActive(_ session: VoiceAudioSession) -> ActiveSession? {
    locked { active?.publicSession.id == session.id ? active : nil }
  }

  private func clearActive(_ session: ActiveSession) {
    locked {
      if active === session {
        active = nil
      }
    }
  }

  private func locked<T>(_ action: () -> T) -> T {
    lock.lock()
    defer { lock.unlock() }
    return action()
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/voice/audio/VoiceAudioHub.kt
package com.galaxyssi.chat.voice.audio

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import java.util.UUID
import java.util.concurrent.atomic.AtomicBoolean

data class VoiceAudioSessionConfig(
    val capture: PcmCaptureConfig = PcmCaptureConfig(),
    val endpoint: AdaptiveEndpointConfig = AdaptiveEndpointConfig(),
    val autoEndpoint: Boolean = false
)

data class VoiceAudioSession(val id: String)

interface VoiceAudioHubListener {
    val acceptsDirectPcmFrames: Boolean
        get() = false

    val acceptsPcmFrames: Boolean
        get() = false

    fun onCaptureReady(session: VoiceAudioSession, state: PcmRecorderState) = Unit
    fun onDirectPcmFrame(session: VoiceAudioSession, frame: DirectPcmFramePacket) = Unit
    fun onPcmFrame(session: VoiceAudioSession, frame: PcmFramePacket) = Unit
    fun onAudioLevel(session: VoiceAudioSession, decision: VadDecision) = Unit
    fun onSpeechStarted(session: VoiceAudioSession, sequence: Long) = Unit
    fun onSpeechEndedCandidate(session: VoiceAudioSession, sequence: Long) = Unit
    fun onEndpoint(session: VoiceAudioSession, reason: EndpointReason) = Unit
    fun onInputRouteChanged(session: VoiceAudioSession, route: String) = Unit
    fun onFailure(session: VoiceAudioSession, error: Throwable) = Unit
}

class VoiceAudioHub(
    private val recorder: PcmRecorder,
    private val scope: CoroutineScope,
    private val sessionIdFactory: () -> String = { UUID.randomUUID().toString() },
    private val vadFactory: () -> VoiceActivityDetector = { AdaptiveSpeechVad() }
) {
    private data class ActiveSession(
        val public: VoiceAudioSession,
        val config: VoiceAudioSessionConfig,
        val store: SpeechSegmentStore,
        val vad: VoiceActivityDetector,
        val endpoint: AdaptiveEndpointDetector,
        val listener: VoiceAudioHubListener,
        val stopRequested: AtomicBoolean = AtomicBoolean(false),
        val completion: CompletableDeferred<Unit> = CompletableDeferred(),
        var lastRoute: String = "",
        var job: Job? = null
    )

    private val lock = Any()
    private var active: ActiveSession? = null

    fun start(
        config: VoiceAudioSessionConfig,
        listener: VoiceAudioHubListener = object : VoiceAudioHubListener {}
    ): VoiceAudioSession? {
        val session = synchronized(lock) {
            if (active != null) return null
            val public = VoiceAudioSession(sessionIdFactory())
            ActiveSession(
                public = public,
                config = config,
                store = InMemorySpeechSegmentStore(config.capture.sampleRateHz, config.capture.maxDurationMs + 1_000L),
                vad = vadFactory(),
                endpoint = AdaptiveEndpointDetector(
                    config.capture.sampleRateHz,
                    config.endpoint,
                    config.autoEndpoint
                ),
                listener = listener
            ).also { active = it }
        }
        session.job = scope.launch {
            try {
                val frames = recorder.start(session.config.capture)
                if (session.stopRequested.get()) {
                    recorder.stop(PcmStopReason.USER_CANCEL)
                    return@launch
                }
                val readyState = recorder.currentState()
                session.lastRoute = readyState.inputRoute
                safely { session.listener.onCaptureReady(session.public, readyState) }
                frames.collect { frame ->
                    try {
                        processFrame(session, frame)
                    } finally {
                        frame.close()
                    }
                }
            } catch (error: Throwable) {
                if (!session.stopRequested.get()) {
                    safely { session.listener.onFailure(session.public, error) }
                }
            } finally {
                session.completion.complete(Unit)
                if (!session.stopRequested.get()) {
                    session.store.clear()
                    synchronized(lock) {
                        if (active === session) active = null
                    }
                }
            }
        }
        return session.public
    }

    fun currentAmplitude(): Int = recorder.currentState().currentAmplitude

    fun currentState(): PcmRecorderState = recorder.currentState()

    fun requestStop(session: VoiceAudioSession, reason: PcmStopReason) {
        val current = synchronized(lock) { active?.takeIf { it.public.id == session.id } } ?: return
        current.stopRequested.set(true)
        recorder.requestStop(reason)
    }

    suspend fun stop(session: VoiceAudioSession, reason: PcmStopReason): VoiceAudioCaptureResult? {
        val current = synchronized(lock) { active?.takeIf { it.public.id == session.id } } ?: return null
        current.stopRequested.set(true)
        recorder.stop(reason)
        current.completion.await()
        val state = recorder.currentState()
        val result = VoiceAudioCaptureResult(
            sessionId = session.id,
            stopReason = state.stopReason ?: reason,
            snapshot = current.store.snapshot(
                SegmentRange(
                    preRollMs = current.config.endpoint.preRollMs,
                    postRollMs = current.config.endpoint.postRollMs
                )
            ),
            diagnostics = state.diagnostics,
            audioSource = state.audioSource,
            inputRoute = state.inputRoute
        )
        current.store.clear()
        synchronized(lock) {
            if (active === current) active = null
        }
        return result
    }

    fun activeSession(): VoiceAudioSession? = synchronized(lock) { active?.public }

    fun snapshotWindow(session: VoiceAudioSession, maxDurationMs: Long): PcmSnapshot? {
        val current = synchronized(lock) { active?.takeIf { it.public.id == session.id } } ?: return null
        return current.store.snapshotWindow(maxDurationMs)
    }

    private fun processFrame(session: ActiveSession, frame: AudioFrame) {
        session.store.append(frame)
        if (session.listener.acceptsDirectPcmFrames) {
            frame.directPcm16Buffer()?.let { pcm16 ->
                safely {
                    session.listener.onDirectPcmFrame(
                        session.public,
                        DirectPcmFramePacket(
                            sequence = frame.sequence,
                            captureTimeNanos = frame.captureTimeNanos,
                            pcm16 = pcm16,
                            sampleCount = frame.validSamples,
                            sampleRateHz = session.config.capture.sampleRateHz
                        )
                    )
                }
            }
        }
        if (session.listener.acceptsPcmFrames) {
            safely {
                session.listener.onPcmFrame(
                    session.public,
                    PcmFramePacket(
                        sequence = frame.sequence,
                        captureTimeNanos = frame.captureTimeNanos,
                        samples = frame.samples.copyOf(frame.validSamples),
                        sampleRateHz = session.config.capture.sampleRateHz
                    )
                )
            }
        }
        val vad = session.vad.accept(frame)
        val endpoint = session.endpoint.accept(frame, vad)
        if (endpoint.speechStarted) {
            session.store.markSpeechStart(frame.sequence)
            safely { session.listener.onSpeechStarted(session.public, frame.sequence) }
        }
        if (endpoint.speechEndedCandidate) {
            session.store.markSpeechEnd(frame.sequence)
            safely { session.listener.onSpeechEndedCandidate(session.public, frame.sequence) }
        }
        safely { session.listener.onAudioLevel(session.public, vad) }
        val route = recorder.currentState().inputRoute
        if (route.isNotBlank() && route != session.lastRoute) {
            session.lastRoute = route
            safely { session.listener.onInputRouteChanged(session.public, route) }
        }
        endpoint.endpointReason?.let { reason ->
            if (session.stopRequested.compareAndSet(false, true)) {
                safely { session.listener.onEndpoint(session.public, reason) }
            }
        }
    }

    private inline fun safely(block: () -> Unit) {
        runCatching(block)
    }
}
